x/gamm/pool-models/stableswap/amm_test.go

```diff
@@ -100,3 +106,104 @@ func TestCFMMInvariantMultiAssets(t *testing.T) {
 		decApproxEq(t, k2, k3, kErrTolerance)
 	}
 }
+
+func (suite *StableSwapTestSuite) Test_StableSwap_CalculateAmountOutAndIn_InverseRelationship(t *testing.T) {
+	type testcase struct {
+		denomOut         string
+		initialPoolOut   int64
+		initialWeightOut int64
+		initialCalcOut   int64
+
+		denomIn         string
+		initialPoolIn   int64
+		initialWeightIn int64
+	}
+
+	// For every test case in testcases, apply a swap fee in swapFeeCases.
+	testcases := []testcase{
+		{
+			denomOut:         "uosmo",
+			initialPoolOut:   1_000_000_000_000,
+			initialWeightOut: 100,
+			initialCalcOut:   100,
+
+			denomIn:         "ion",
+			initialPoolIn:   1_000_000_000_000,
+			initialWeightIn: 100,
+		},
+		{
+			denomOut:         "uosmo",
+			initialPoolOut:   1_000,
+			initialWeightOut: 100,
+			initialCalcOut:   100,
+
+			denomIn:         "ion",
+			initialPoolIn:   1_000_000,
+			initialWeightIn: 100,
+		},
+		{
+			denomOut:         "uosmo",
+			initialPoolOut:   1_000,
+			initialWeightOut: 100,
+			initialCalcOut:   100,
+
+			denomIn:         "ion",
+			initialPoolIn:   1_000_000,
+			initialWeightIn: 100,
+		},
+		{
+			denomOut:         "uosmo",
+			initialPoolOut:   1_000,
+			initialWeightOut: 200,
+			initialCalcOut:   100,
+
+			denomIn:         "ion",
+			initialPoolIn:   1_000_000,
+			initialWeightIn: 50,
+		},
+		{
+			denomOut:         "uosmo",
+			initialPoolOut:   1_000_000,
+			initialWeightOut: 200,
+			initialCalcOut:   100000,
+
+			denomIn:         "ion",
+			initialPoolIn:   1_000_000_000,
+			initialWeightIn: 50,
+		},
+	}
+
+	swapFeeCases := []string{"0", "0.001", "0.1", "0.5", "0.99"}
+
+	getTestCaseName := func(tc testcase, swapFeeCase string) string {
+		return fmt.Sprintf("tokenOutInitial: %d, tokenInInitial: %d, initialOut: %d, swapFee: %s",
+			tc.initialPoolOut,
+			tc.initialPoolIn,
+			tc.initialCalcOut,
+			swapFeeCase,
+		)
+	}
+
+	for _, tc := range testcases {
+		for _, swapFee := range swapFeeCases {
+			t.Run(getTestCaseName(tc, swapFee), func(t *testing.T) {
+				ctx := createTestContext(t)
+
+				poolLiquidityIn := sdk.NewInt64Coin(tc.denomOut, tc.initialPoolOut)
+				poolLiquidityOut := sdk.NewInt64Coin(tc.denomIn, tc.initialPoolIn)
+				poolLiquidity := sdk.NewCoins(poolLiquidityIn, poolLiquidityOut)
+
+				swapFeeDec, err := sdk.NewDecFromStr(swapFee)
+				require.NoError(t, err)
+
+				exitFeeDec, err := sdk.NewDecFromStr("0")
+				require.NoError(t, err)
+
+				pool := createTestPool(t, poolLiquidity, swapFeeDec, exitFeeDec)
+				require.NotNil(t, pool)
```

Comment:
I see that all this setup is duplicated with the balancer pool. Can we also move everything on lines 190-203 to the common test suite `CfmmCommonTestSuite`?

Reply:
For this comment, seems to me that moving line 190-203 as a separate method would cause an extra overhead, since there's not much logic here and the method would simply return what it got as a parameter, when we can just assign values directly. Which would look like this
```
func (suite *CfmmCommonTestSuite) preparePoolVariables(
      denomOut string,
      initialPoolOut int64,
      denomIn string, 
      initialPoolIn string,
      swapFee string,
      exitFee string)(poolLiquidityIn sdk.Coin, poolLiquidityOut sdk.Coin, swapFeeDec sdk.Dec, exitFeeDec sdk.Dec) {
}
```